File: USACO_training/ariprog.cpp
/*
ID: s.sophi1
LANG: C++14 
PROG: ariprog               
*/

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <set>
#include <utility> 

using namespace std;

int search_length;

set<pair<int, int>> try_all_possibilities(set<int> bisquares, bool bisquares_flags[]) {
	set<pair<int, int>> a_and_bs;
    vector<int> bisquares_list(bisquares.begin(), bisquares.end());
    sort(bisquares_list.begin(), bisquares_list.end());
    
    for (int i = 0; i <= bisquares.size()-search_length; i++) {
    	int a = bisquares_list[i];
    	int num = bisquares_list.back() - bisquares_list[0];
    	for (int j = 1; j <= (num-a)/(search_length-1); j++) {
    		int b = j;
    		bool bad = false;
    		for (int x=search_length-1; x>=0; x--) {
    			if (!bisquares_flags[a + (x*b)]) {
    				bad = true;
    				break;
    			} 
    		}

    		if (!bad)
    			a_and_bs.insert({b, a});
    	}
    } 
      
    return a_and_bs;
}


int main() {
	ifstream fin ("ariprog.in");
    ofstream fout ("ariprog.out");
    
    int pq_limit;
    fin >> search_length >> pq_limit;

    bool bisquares_flags[2*pq_limit*pq_limit] {false};


    set<int> bisquare_set;
    for (int p = 0; p <= pq_limit; p++) {
        for (int q = p; q <= pq_limit; q++) {
            int bisquare = p*p + q*q;
            bisquares_flags[bisquare] = true;
            bisquare_set.insert(bisquare);
        }
    }

    auto thing = try_all_possibilities(bisquare_set, bisquares_flags);
    if (!thing.empty()) {
    	for (const auto & p: thing)
    		fout << p.second << " " << p.first << endl;
    }	
    else 
    	fout << "NONE" << endl;
    return 0;
}

File: USACO_training/ariprog.py
"""
ID: s.sophi1
LANG: PYTHON3
PROG: ariprog
"""

input = open('ariprog.in', 'r')
output = open('ariprog.out', 'w')
strings = input.readlines()
strings = [string.rstrip() for string in strings]
search_length = int(strings[0])
pq_limit = int(strings[1])
input.close()

bisquare_set = set()
for p in range(pq_limit+1):
    for q in range(p, pq_limit+1):
        bisquare_set.add(p*p+q*q)
print(bisquare_set)


def try_all_possibilities(bisquares):
    a_and_bs = set()
    bisquares_list = sorted(bisquares)
    for i in range(0, len(bisquares)-search_length+1):
        a = bisquares_list[i]
        num = bisquares_list[-1]-bisquares_list[0]
        for j in range(1, int((num-a)/(search_length-1)+1)):
            b = j
            bad = False
            for x in range(search_length-1, 0, -1):
                if a + (x*b) not in bisquares:
                    bad = True
                    break
            if not bad:
                a_and_bs.add((a, b))
    return a_and_bs


thing = sorted(try_all_possibilities(bisquare_set), key = lambda x: (x[1], x[0]))
if thing:
    for item in thing:
        output.write(str(item[0]) + ' ' + str(item[1]) + '\n')
else:
    output.write('NONE' + '\n')
output.close()
